### ISSUE TITLE
Proposal: Improve Port Mapping Syntax

### ISSUE THREAD

**Issue by dave-tucker** [docker/compose#2605]
In the current Docker CLI (as of 1.9) there are at least 4 ways of specifying a port mapping for `-p`
- `containerPort`
- `hostPort:containerPort`
- `hostIP:hostPort:containerPort`
- `hostIP::containerPort`

As Compose is looking to make some incompatible changes in the v2 format I would like to propose that we move away from using the `string` and to instead use a `hash`/`dictionary`

**v1**

```
ports:
  - "80"
  - "8080:80"
  - "192.168.99.100:8080:80"
  - "192.168.99.100::80"
```
- Proposed v2**

```
ports:
  - container_port: 80

  - host_port: 8080
    container_port: 80

  - host_ip: "192.168.99.100"
    host_port: 8080
    container_port: 80

  - host_ip: "192.168.99.100"
    container_port: 80
```

The new format should be:
- Easier to read
- Easier to validate as the port numbers are `int` rather than `string` and there are no separators to deal with
- Easier to document - we replace 4 different string formats with 3 keys

`container_port` is required, all other fields are optional
`host_port` defines a specific port on the host
`host_ip` specifies which IP address a port should bind to (default 0.0.0.0)

**Comment by aanand** [comment]
Linked to from https://github.com/docker/compose/issues/2504

**Comment by dnephin** [comment]
I believe there are also port ranges that can be used as both host and container port.

Repeating `port` in the keys in a `ports` section seems redundant. Could we do:

```yaml

ports:
   - host: 80-90
     container: 8080-8090
     host_ip: 192.168.99.190
```

I expect there will be users who are unhappy with this change, especially if they have many ports in their compose file.

I'd like to see us keep the minimal form for the first two string formats `80` and `80:8000`. If you need an IP or range, then you must use the mapping form. That way most users can continue to use the less-verbose, and really not difficult to parse version, and the few users who actually need to specify a host ip or port range can use the more verbose option.

**Comment by dave-tucker** [comment]
@dnephin good catch on the port ranges and :+1: on dropping "port" from the keys.

IMO being a little more verbose actually helps readability/maintainability so I'd be in favour of dropping the strings entirely. It's quite easy to forget what "8080:80" means and in the case of `hostPort:containerPort` and we're only adding one additional line of config! Not to mention, it's much easier for new or inexperienced users to understand.

I'd also love to see something that can take one or more Docker CLI commands, and spit out a Compose v2 file. Either a web app (that is part of the docs) and maybe a CLI command. I know it's outside the scope of this issue, but it my be cool to be able to do something like this:

```
$ echo "-it --net=host -p "8080:80" foo" | docker-compose servicify myapp 
myapp:
     tty: true
     network: host
     ports:
       - host: 8080
         container: 80
```

**Comment by aluzzardi** [comment]
:+1:

**Comment by railnet** [comment]
My question here regards the following scenario.
I suppose to have a service attached to two (or more) networks, as configured here below:
    networks:
      front:
        ipv4_address: 172.18.0.1
      back:
        ipv4_address: 170.20.0.1

and let suppose that service has to expose some ports too:
    ports:
     - "4189:9999"

How I can grant the iptables configured by Docker at the host (in order to implement the port mapping at the host) will map the host_port 4189 to the container IPv4 172.18.0.1?
It seems the choice to use 172.18.0.1 or 170.20.0.1 is ramdom at today.
Ref also 
https://github.com/docker/docker/issues/21741
I also inserted a proposal  there.
Many Thanks in advance for your suggestions.

**Comment by shin-** [comment]
The modern version of this proposal was implemented in format 3.2 - closing.
